### PR TITLE
Add support for C# local-debugging

### PR DESCRIPTION
**Pull request by nima-ap-zz**
Implement local debugging in C# sdk

### CODE REVIEW

File: targets/csharp/make.js [hunk 1]
```diff
@@ -337,6 +338,32 @@ function getRequestActions(tabbing, apiCall, isInstance) {
     return "";
 }
 
+function getCustomApiLogic(tabbing, apiCall) {
+    if (apiCall.name === "ExecuteFunction")
+    {
+        return "\n" + tabbing + "string localApiServerString = PlayFabSettings.LocalApiServer;\n"
+            + tabbing + "var debugFunctions = PlayFabSettings.LocalDebugFunctions;\n"
+            + tabbing + "var shouldDebugFunction = debugFunctions.Contains(\"*\") || debugFunctions.Contains(request.FunctionName);\n"
```
Comment (MGudgin) [inline targets/csharp/make.js]: Equals rather than Contains for the first clause. Plus see comments on the Unity SDK as to whether we should do the specific function feature or not.

File: targets/csharp/source/source/PlayFabHttp/PlayFabHttp.cs [hunk 2]
```diff
@@ -50,6 +50,17 @@ public static async Task<object> DoPost(string urlPath, PlayFabRequestCommon req
...
             var fullPath = settings.GetFullUrl(urlPath);
+            return await _DoPost(fullPath, request, authType, authKey, extraHeaders, instanceSettings);
+        }
+
+        public static async Task<object> DoPostWithFullUri(string fullUriPath, PlayFabRequestCommon request, string authType, string authKey, Dictionary<string, string> extraHeaders, PlayFabApiSettings instanceSettings = null)
+        {
+            return await _DoPost(fullUriPath, request, authType, authKey, extraHeaders, instanceSettings);
+        }
+
+        public static async Task<object> _DoPost(string fullPath, PlayFabRequestCommon request, string authType, string authKey, Dictionary<string, string> extraHeaders, PlayFabApiSettings instanceSettings = null)
```
Comment (MGudgin) [inline targets/csharp/source/source/PlayFabHttp/PlayFabHttp.cs]: I suspect this one should be private

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 3]
```diff
@@ -49,6 +51,100 @@ namespace PlayFab
...
+        private static StringBuilder _sb;
+        /// <summary>
+        /// A threadsafe way to block and load a text file
+        /// 
+        /// Load a text file, and return the file as text.
+        /// Used for small (usually json) files.
+        /// </summary>
+        public static string ReadAllFileText(string filename)
+        {
+            if (!File.Exists(filename))
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs:65]: Here and elsewhere, please use braces even for single statement blocks.

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 4]
```diff
@@ -49,6 +51,100 @@ namespace PlayFab
...
+                return string.Empty;
+                
+            if (_sb == null)
+                _sb = new StringBuilder();
+            _sb.Length = 0;
+
+            FileStream fs = null;
+            try
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs]: I don't understand why we have the try/finally thing rather than just having two using blocks, one for FileStream and one for BinaryReader.

File: targets/unity-v2/make.js [hunk 5]
```diff
@@ -484,7 +484,7 @@ function getCustomApiLogic(tabbing, apiCall)
 {
     if (apiCall.name === "ExecuteFunction")
         return tabbing + "var localApiServerString = PlayFabSettings.LocalApiServer;\n" +
-            tabbing + "if (!string.IsNullOrEmpty(localApiServerString))\n" +
+            tabbing + "if (!string.IsNullOrEmpty(localApiServerString) && PlayFab.Internal.PlayFabUtil.ShouldDebugFunction(request.FunctionName))\n" +
```
Comment (MGudgin) [inline targets/unity-v2/make.js]: Should this be in a separate PR for the Unity SDK?

File: targets/unity-v2/source/Shared/Internal/Util.cs [hunk 6]
```diff
@@ -144,6 +144,12 @@ internal static string GetLocalSettingsFileProperty(string propertyKey)
             }
             return string.Empty;
         }
+
```
Comment (MGudgin) [inline targets/unity-v2/source/Shared/Internal/Util.cs]: Should this be in a separate PR for the Unity SDK?

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 7]
```diff
@@ -49,6 +51,100 @@ namespace PlayFab
...
+                }
+                catch (KeyNotFoundException)
+                {
+                    return string.Empty;
+                }
+            }
+            return string.Empty;
+        }
+
+        internal static List<string> GetLocalDebugFunctions()
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs]: It feels odd having GetLocalSettingsFileProperty, which is general and GetLocalDebugFunctions which is specific. I'd expect withe GetLocalDebugUri/GetLocalDebugFunctions or GetLocalSettingsFileProperty/GetLocalSettingsFileArrayProperty.

File: targets/csharp/make.js (hunk 1 above)
Comment (MGudgin) [reply targets/csharp/make.js]: See my comment on https://github.com/nima-ap/SDKGenerator/pull/2

File: targets/csharp/templates/PlayFabUtil.cs.ejs (hunk 3 above)
Comment (nima-ap) [reply targets/csharp/templates/PlayFabUtil.cs.ejs]: This was not my code, it's just me reviving code they had and got rid of. I'll fix it.

File: targets/csharp/templates/PlayFabUtil.cs.ejs (hunk 4 above)
Comment (nima-ap-zz) [reply targets/csharp/templates/PlayFabUtil.cs.ejs]: Above comment for braces answers this. I'll update it with our using pattern.

File: targets/unity-v2/make.js (hunk 5 above)
Comment (nima-ap-zz) [reply targets/unity-v2/make.js]: Already exists in Unity SDK PR I believe; it somehow bled into this. Removing on next commit.

File: targets/csharp/templates/PlayFabUtil.cs.ejs (hunk 3 above)
Comment (nima-ap-zz) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:65]: Fixing this shortly.

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 8]
```diff
@@ -64,16 +64,18 @@ namespace PlayFab
...
             _sb.Length = 0;
 
-            FileStream fs = null;
-            try 
+            using (var fs = new FileStream(filename, FileMode.Open))
             {
-                fs = new FileStream(filename, FileMode.Open);
                 using (BinaryReader br = new BinaryReader(fs))
                 {
                     // Attempt to release the descriptor immediately
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs]: Given the using, this is probably a bad idea, no?

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 9]
```diff
@@ -83,12 +85,6 @@ namespace PlayFab
                         _sb.Append(br.ReadChar());
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs:82]: braces

**Review by MGudgin — CHANGES_REQUESTED** [review]

File: targets/csharp/make.js (hunk 1 above)
Comment (nima-ap-zz) [reply targets/csharp/make.js]: Please either explicitly request removal of partial debugging or resolve this to proceed with PR.

Comment (MGudgin) [reply targets/csharp/make.js]: OK. Please remove partial debugging from this and related PRs. Please put the partial debugging code into a separate branch or branches to allow us to merge it later if we decide to do so.

Comment (nima-ap-zz) [reply targets/csharp/make.js]: Removed.

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 10]
```diff
@@ -49,6 +51,82 @@ namespace PlayFab
...
+                    while (br.BaseStream.Position != br.BaseStream.Length)
+                    {
+                        _sb.Append(br.ReadChar());
+                    }
+                }
+            }
+            return _sb.ToString();
+        }
+
+        internal static string GetLocalSettingsFileProperty(string propertyKey)
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs:89]: Can this be private instead of internal? Or do we need it to be internal for testing purposes?

Comment (nima-ap-zz) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: It needs to be accessed by `PlayFabSettings.cs`

Comment (MGudgin) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: OK. Should ReadAllFileText be private then?

Comment (nima-ap-zz) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: Unfortunately I believe it cannot be. 
C:\github\PlayFab\Fork\SDKGenerator\targets\unity-v2\Testing\Tests\Shared\TestTitleDataLoader.cs(49):                    testTitleDataContents = PlayFabUtil.ReadAllFileText(filename);

Comment (MGudgin) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: OK. That makes sense. Did you add the test that calls ReadAllFileText? 

What I'm concerned about is that we have two methods here, both of which are called from outside the class, but both have different access modifiers. I don't understand why they are not the same (either both public or both internal). If one is only called by test code, but the other is called by customer code, then I could understand them being different as in that case the one called by customer code would be public and the test one would (or at least could) be internal (assuming InternalsVisibleTo attribute exists somewhere).

Comment (nima-ap-zz) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: Actually I apologize; this method is not being used by the C# SDK. That search query returned a result for a **unity** SDK test call to `ReadAllFileText`. This method can indeed be private as it is being used by code in `PlayFabUtil.cs` only.

The reason for all this confusion is the fact that the SDK team created this method for the C# SDK before we even started working on local-debugging. However, during our work on it, they decided to remove the method as it was no longer needed since they changed their approach for getting test title data on the C# SDK due to all the complexities of doing File IO when targetting 4 frameworks that not all support doing it (portable4.5).

It is due to this fact that the earlier version of the code you saw in this looked completely different in terms of coding style, such as the lack of `using` and `{` which is a pattern I've gotten used to while working here.

In short, we decided to bring this method back as we **need** it for our work but at the same time decided to `#if` define it because we don't expect people to want to run local-debugging on a portable 4.5 device.

So yes, this method can be private, and it will be in my commit shortly.

Comment (MGudgin) [reply targets/csharp/templates/PlayFabUtil.cs.ejs:89]: Cool. Thanks for reminding me how we got here.

**Review by MGudgin — APPROVED** [review]

File: targets/csharp/templates/PlayFabUtil.cs.ejs [hunk 11]
```diff
@@ -3,6 +3,7 @@ using System;
 using System.IO;
 using System.Text;
 using System.Threading.Tasks;
+using System.Collections.Generic;
```
Comment (MGudgin) [inline targets/csharp/templates/PlayFabUtil.cs.ejs]: Remove this line.

**Review by MGudgin — APPROVED** [review]

**Comment by nima-ap-zz** [comment]
Closing PR in favor of the new one opened that follows an improved branching structure.
